### PR TITLE
modPow: support negative base and exponent

### PR DESCRIPTION
**Pull request by dragonsinth**

### CODE REVIEW

**Comment by dragonsinth** [comment]
@thhofer please take al ook at this version

**Comment by thhofer** [comment]
Looks good! Definitely an improvement on my suggestion.

The only issue I might have with it: some operations in the secure version now depend on the sign of the exponent. Thus, timing attacks might reveal the sign of the exponent, which might or might not be an issue depending on context, but that's precisely what we'd like to avoid in the secure version, right?

**Comment by dragonsinth** [comment]
I don't think it should be a big problem, the cost of the modPow should dwarf everything else in any cases where it could matter.  @JakeWharton any thoughts?

File: jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java [hunk 1]
```diff
@@ -267,10 +268,22 @@ private BigInteger modPowInsecureImpl(BigInteger base, BigInteger exp, BigIntege
...
+    }
+
     mpz_t basePeer = getPeer(base, sharedOperands[0]);
     mpz_t expPeer = getPeer(exp, sharedOperands[1]);
     mpz_t modPeer = getPeer(mod, sharedOperands[2]);
 
+    if (invert) {
+      int res = __gmpz_invert(basePeer, basePeer, modPeer);
+      if (res == 0) {
+        throw new ArithmeticException("val not invertible");
```
Comment (JakeWharton) [inline jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java]: Can we test for this case?

File: jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java [hunk 2]
```diff
@@ -255,10 +244,22 @@ private int kroneckerImpl(BigInteger a, BigInteger p) {
...
```
Comment (JakeWharton) [inline jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java]: Can we test for this case?

**Comment by JakeWharton** [comment]
I don't know enough to have an opinion. I can try to find someone that would if it's needed.

File: jnagmp/src/test/java/com/squareup/jnagmp/GmpTest.java [hunk 3]
```diff
@@ -364,8 +334,8 @@ private void testEvenExamples() {
   }
 
   @Test public void testSmallExhaustiveSecure() {
-    for (int base = 10; base >= 0; --base) {
-      for (int exp = 10; exp >= 0; --exp) {
+    for (int base = 10; base >= -10; --base) {
+      for (int exp = 10; exp >= -10; --exp) {
```
Comment (dragonsinth) [inline jnagmp/src/test/java/com/squareup/jnagmp/GmpTest.java]: @JakeWharton those cases are tested here and 311.  We're testing that the GMP stuff either returns the same result or else throws the same exception type as Java BigInteger

Comment (JakeWharton) [inline jnagmp/src/test/java/com/squareup/jnagmp/GmpTest.java]: Got it. Yeah I didn't spot that in review but I see it now.

**Pull request merged**